grr/server/grr_response_server/bin/grrafana.py

```diff
@@ -0,0 +1,165 @@
+import os
+import json
+import typing
+from typing import Any, List, Tuple, Dict, Text
+
+from werkzeug.wrappers import Request, Response
+from werkzeug.wrappers.json import JSONMixin
+from werkzeug.routing import Map, Rule
+from werkzeug.exceptions import HTTPException, NotFound
+from werkzeug.middleware.shared_data import SharedDataMiddleware
+from werkzeug.utils import redirect
+from werkzeug.serving import run_simple
+
+from absl import app
+from absl import flags
+
+from grr_response_core.config import server as config_server
+from grr_response_server import fleetspeak_connector
+from grr_response_server import fleetspeak_utils
+from grr_response_server import server_startup
+from grr_response_core import config
+from grr_response_core.config import contexts
+
+from fleetspeak.src.server.proto.fleetspeak_server.resource_pb2 import ClientResourceUsageRecord
+
+
+AVAILABLE_METRICS = ['mean_user_cpu_rate', 'max_user_cpu_rate', 'mean_system_cpu_rate', 
+                    'max_system_cpu_rate', 'mean_resident_memory_mib', 'max_resident_memory_mib']
+
+flags.DEFINE_bool(
+  "version",
+  default=False,
+  allow_override=True,
+  help="Print the GRR console version number and exit immediately.")
+
+
+class JSONRequest(JSONMixin, Request):
+  pass
+
+
+class JSONResponse(JSONMixin, Response):
+  pass
+
+
+class Grrafana(object):
+  """GRRafana HTTP server instance.
+  
+  A full description of all endpoints implemented within this HTTP
+  server can be found in https://github.com/simPod/grafana-json-datasource#api."""
+
+  def __init__(self, config: dict) -> None:
+    """Constructor."""
+    self.url_map = Map([
+    Rule('/', endpoint='Root', methods=["GET"]),
+    Rule('/search', endpoint='Search', methods=["POST"]),
+    Rule('/query', endpoint='Query', methods=["POST"]),
+    Rule('/annotations', endpoint='Annotations', methods=["POST"])
+    ])
+
+  def dispatchRequest(self, request: JSONRequest) -> Any:
+    """Maps requests to different methods."""
+    adapter = self.url_map.bind_to_environ(request.environ)
+    try:
+      endpoint, values = adapter.match()
+      return getattr(self, 'on' + endpoint)(request, **values)
+    except HTTPException as e:
+      return e
+
+  def wsgiApp(self, environ, start_response) -> Response:
+    request = JSONRequest(environ)
+    response = self.dispatchRequest(request)
+    return response(environ, start_response)
+
+  def __call__(self, environ, start_response):
+    return self.wsgiApp(environ, start_response)
+
+  def onRoot(self, request: JSONRequest) -> JSONResponse:
+    """Returns OK message to database connection check."""
+    return JSONResponse()
+
+  def onSearch(self, request: JSONRequest) -> JSONResponse:
+    """Depending on the request type, returns either available client
+    resource usage metrics from Fleetspeak database, or possible values
+    for a defined Grafana variable (currently supports only variables based
+    on client IDs)."""
+    if "type" in request.json:
```

Comment:
As far as I understand presense of "type" in request.json means different Grafana request types. Please add a comment detailing this.